File: umbra-go/sync_test.go
package umbra

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestSyncExchangeIsSignedAndReturnsConflictsAsData(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/api/v1/client/sync/exchange", func(w http.ResponseWriter, r *http.Request) {
		var input SyncExchangeInput
		raw := decodeJSONBody(t, r, &input)
		assertSignedRequest(t, r, raw, "device-1", "device-secret")
		if input.Space.Name != "library" || len(input.Mutations) != 1 {
			t.Fatalf("unexpected input: %+v", input)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{
			"code": 0, "msg": "success",
			"data": SyncExchangeResult{
				Accepted: []SyncAcceptedMutation{}, Rejected: []SyncRejectedMutation{}, Changes: []SyncChange{},
				Conflicts:  []SyncConflict{{MutationID: "m-1", Reason: "base_version_mismatch"}},
				NextCursor: "cursor-1",
			},
		})
	})
	server := httptest.NewServer(mux)
	defer server.Close()
	client := newSignedTestClient(t, server.URL, "device-1", "device-secret")
	mutation, err := NewUpsertMutation("m-1", SyncRecordKey{Namespace: "lunabox.library", Collection: "games", RecordID: "game-1"}, 1, 2, map[string]any{"name": "Example"})
	if err != nil {
		t.Fatal(err)
	}
	result, err := client.Sync.Exchange(context.Background(), SyncExchangeInput{Space: SyncSpace{Name: "library"}, Mutations: []SyncMutation{mutation}})
	if err != nil {
		t.Fatalf("Exchange() error = %v", err)
	}
	if len(result.Conflicts) != 1 || result.Conflicts[0].MutationID != "m-1" {
		t.Fatalf("unexpected result: %+v", result)
	}
}

func TestSyncSnapshotQueryIsSigned(t *testing.T) {
	mux := http.NewServeMux()
	mux.HandleFunc("/api/v1/client/sync/snapshot", func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Query().Get("protocol_version") != "1" || r.URL.Query().Get("space") != "library" || r.URL.Query().Get("cursor") != "page-1" || r.URL.Query().Get("limit") != "25" {
			t.Fatalf("unexpected query: %s", r.URL.RawQuery)
		}
		assertSignedRequest(t, r, nil, "device-1", "device-secret")
		_ = json.NewEncoder(w).Encode(map[string]any{
			"code": 0, "msg": "success", "data": SyncSnapshotPage{Records: []SyncChange{}, ExchangeCursor: "cursor-2"},
		})
	})
	server := httptest.NewServer(mux)
	defer server.Close()
	client := newSignedTestClient(t, server.URL, "device-1", "device-secret")
	result, err := client.Sync.Snapshot(context.Background(), SyncSnapshotInput{SpaceName: "library", Cursor: "page-1", Limit: 25})
	if err != nil {
		t.Fatalf("Snapshot() error = %v", err)
	}
	if result.ExchangeCursor != "cursor-2" {
		t.Fatalf("unexpected result: %+v", result)
	}
}

func TestSyncChangeDecodePayload(t *testing.T) {
	change := SyncChange{Payload: json.RawMessage(`{"name":"Example"}`)}
	var payload struct {
		Name string `json:"name"`
	}
	if err := change.DecodePayload(&payload); err != nil || payload.Name != "Example" {
		t.Fatalf("DecodePayload() = %+v, %v", payload, err)
	}
}

File: umbra-typescript/tests/sync.test.ts
import { afterEach, describe, expect, it } from 'vitest'
import type { IncomingHttpHeaders } from 'node:http'
import {
  createDeviceSignature,
  MemoryDeviceCredentialStore,
  MemoryTokenStore,
  newUpsertMutation,
  UmbraClient,
} from '../src'
import { json, readBody, startServer } from './helpers'

const servers: Array<{ close: () => Promise<void> }> = []

afterEach(async () => {
  await Promise.all(servers.splice(0).map(server => server.close()))
})

describe('structured sync', () => {
  it('signs exchange and returns conflicts as result data', async () => {
    const client = await signedClient(async (req, res) => {
      if (req.method === 'POST' && req.url === '/api/v1/client/sync/exchange') {
        const body = await readBody(req)
        const parsed = JSON.parse(body)
        expect(parsed.protocol_version).toBe(1)
        expect(parsed.space).toEqual({ name: 'library' })
        await expectSignature(req, body)
        json(res, 200, {
          code: 0,
          msg: 'success',
          data: {
            accepted: [],
            conflicts: [{ mutation_id: 'm-1', reason: 'base_version_mismatch' }],
            rejected: [],
            changes: [],
            next_cursor: 'cursor-1',
            has_more: false,
            reset_required: false,
          },
        })
        return
      }
      json(res, 404, { code: 1001, msg: 'not found', data: null })
    })

    const result = await client.sync.exchange({
      space: { name: 'library' },
      mutations: [newUpsertMutation(
        'm-1',
        { namespace: 'lunabox.library', collection: 'games', record_id: 'game-1' },
        1,
        1,
        { name: 'Example' },
      )],
    })
    expect(result.conflicts).toEqual([{ mutation_id: 'm-1', reason: 'base_version_mismatch' }])
  })

  it('signs the canonical snapshot query', async () => {
    const client = await signedClient(async (req, res) => {
      if (req.method === 'GET' && req.url === '/api/v1/client/sync/snapshot?protocol_version=1&space=library&cursor=page-1&limit=25') {
        const body = await readBody(req)
        expect(body).toBe('')
        await expectSignature(req, body)
        json(res, 200, {
          code: 0,
          msg: 'success',
          data: { records: [], exchange_cursor: 'cursor-2', has_more: false },
        })
        return
      }
      json(res, 404, { code: 1001, msg: 'not found', data: null })
    })

    await expect(client.sync.snapshot({ spaceName: 'library', cursor: 'page-1', limit: 25 })).resolves.toEqual({
      records: [],
      exchange_cursor: 'cursor-2',
      has_more: false,
    })
  })
})

async function signedClient(handler: Parameters<typeof startServer>[0]) {
  const tokenStore = new MemoryTokenStore()
  await tokenStore.save({ accessToken: 'token', tokenType: 'bearer', expiresAt: new Date(Date.now() + 3600_000).toISOString() })
  const server = await startServer(handler)
  servers.push(server)
  return new UmbraClient({
    baseUrl: server.url,
    clientId: 'client',
    tokenStore,
    deviceStore: new MemoryDeviceCredentialStore({ deviceId: 'device-1', deviceSecret: 'device-secret' }),
  })
}

async function expectSignature(
  req: { method?: string | undefined, url?: string | undefined, headers: IncomingHttpHeaders },
  body: string,
) {
  expect(req.headers['x-umbra-device-id']).toBe('device-1')
  const timestamp = Number(req.headers['x-umbra-timestamp'])
  const nonce = String(req.headers['x-umbra-nonce'] || '')
  const expected = await createDeviceSignature({
    method: req.method || '',
    pathWithQuery: req.url || '',
    timestamp,
    nonce,
    body,
    deviceId: 'device-1',
    secret: 'device-secret',
  })
  expect(req.headers['x-umbra-body-sha256']).toBe(expected.bodyHash)
  expect(req.headers['x-umbra-signature']).toBe(`v1=${expected.signature}`)
}
